bin/query-cloudwatch

```diff
@@ -33,41 +34,163 @@ class QueryCloudwatch
+      @stdout = stdout
+    end
+
+    def abort; end
+
+    def start; end
+    
+    def handle_row(row)
+      stdout.puts e.values.to_csv
+    end
+    
+    def finish; end
+  end
+
+  class SqliteOutput
+    attr_reader :db, :filename, :stderr
+
+
+    def initialize(filename:, stderr:)
+      @filename = filename
+      @stderr = stderr
+    end
+
+    def abort
+      if db
+        db.rollback if db.transaction_active?
+        db.interrupt
+        db.close
+        @db = nil
+      end
+    end
+
+    def start
+      @db = SQLite3::Database.new(filename)
+      create_tables_if_needed
+      db.transaction
+    end
+
+    def handle_row(row)
+      message = row['@message']
+      raise "Query must include @message in output when using --sqlite" unless message
+
+      timestamp = row['@timestamp']
+      raise "Query must include @timestamp in output when using --sqlite" unless timestamp
+
+      json = JSON.parse(message)
+      
+      success = json.dig('properties', 'event_properties', 'success')
+      success = success ? 1 : (success.nil? ? nil : 0)
+
+      insert_statement.execute(
+        # NOTE: Order matters here
+        {
+          id: json['id'],
+          timestamp:,
+          name: json['name'],
+          uuid: json.dig('properties', 'user_id'),
+          success:,
+          message:,
+        }.values
+      )
+    end
+
+    def finish
+      if db
+        db.commit
+        stderr.puts <<~END
+          Wrote #{db.total_changes.inspect} rows to the 'events' table in #{filename}
+        END
+        db.close
+      end
+
+      @db = nil    
+    end
+
+    def insert_statement
+      @insert_statement ||= db.prepare(
+        <<~SQL
+          INSERT INTO 
+            events (id, timestamp, name, uuid, success, message) 
+            VALUES (?, ?, ?, ?, ?, ?)
+            ON CONFLICT (id) DO NOTHING
+        SQL
+      )
+    end
+
+    def create_tables_if_needed
+      db.execute_batch(
+        <<~SQL
+          CREATE TABLE IF NOT EXISTS events (
+            id TEXT PRIMARY KEY NOT NULL,
+            timestamp TEXT NOT NULL,
+            name TEXT NOT NULL,
+            uuid TEXT NULL,
+            success INTEGER NULL,
+            message TEXT NOT NULL
+          );
+        SQL
+      )
+    end
+  end
+
   attr_reader :config
 
   def initialize(config)
     @config = config
   end
 
-  def run(stdout: STDOUT)
+  def run(stdout: STDOUT, stderr: STDERR)
     if config.count_distinct
       values = {}
       fetch { |row| values[row[config.count_distinct]] = true }
       stdout.puts values.size
-    else
-      fetch { |row| stdout.puts format_response(row) }
+      return
     end
+
+    output = create_output(stdout:, stderr:)
+
+    output.start
+    fetch { |row| output.handle_row(row) }
+    output.finish
+
   rescue Interrupt
     # catch interrupts (ctrl-c) and directly exit, this skips printing the backtrace
+    output.abort
     exit 1
+  rescue
+    output&.abort
+    raise
   end
```

Comment:
I think we could do the same thing with just

```ruby
ensure
  output&.abort
```
which gets called after all `rescue` blocks anyways? maybe set `output = nil` after the finish?